File: gcp_cost_optimizer_agent/go/tools/assets.go
package tools

import (
	"fmt"
	"sort"

	asset "cloud.google.com/go/asset/apiv1"
	"cloud.google.com/go/asset/apiv1/assetpb"
	"google.golang.org/adk/tool"
	"google.golang.org/adk/tool/functiontool"
	"google.golang.org/api/iterator"
)

type ListResourcesArgs struct {
	ProjectID  string   `json:"project_id" description:"The GCP project ID (e.g. 'my-project')."`
	AssetTypes []string `json:"asset_types,omitempty" description:"Optional list of asset types to filter by, e.g. ['compute.googleapis.com/Instance', 'storage.googleapis.com/Bucket']. If omitted, returns all resource types."`
}

type AssetSummaryItem struct {
	AssetType string `json:"asset_type"`
	Count     int    `json:"count"`
}

type ListResourcesResult struct {
	Total   int                 `json:"total"`
	ByType  map[string][]string `json:"by_type"`
	Summary []AssetSummaryItem  `json:"summary"`
}

func NewListResourcesTool() (tool.Tool, error) {
	return functiontool.New(functiontool.Config{
		Name:        "list_resources",
		Description: "List all GCP resources in a project via Cloud Asset Inventory, grouped by type. This is your primary discovery tool — it shows everything deployed in the project.",
	}, ListResources)
}

func ListResources(ctx tool.Context, args ListResourcesArgs) (ListResourcesResult, error) {
	c, err := asset.NewClient(ctx)
	if err != nil {
		return ListResourcesResult{}, fmt.Errorf("creating asset client: %w", err)
	}
	defer c.Close()

	req := &assetpb.ListAssetsRequest{
		Parent:      fmt.Sprintf("projects/%s", args.ProjectID),
		AssetTypes:  args.AssetTypes,
		ContentType: assetpb.ContentType_RESOURCE,
		PageSize:    1000,
	}

	byType := make(map[string][]string)
	it := c.ListAssets(ctx, req)
	for {
		resp, err := it.Next()
		if err == iterator.Done {
			break
		}
		if err != nil {
			return ListResourcesResult{}, fmt.Errorf("iterating assets: %w", err)
		}
		byType[resp.AssetType] = append(byType[resp.AssetType], resp.Name)
	}

	var summary []AssetSummaryItem
	total := 0
	for t, names := range byType {
		summary = append(summary, AssetSummaryItem{
			AssetType: t,
			Count:     len(names),
		})
		total += len(names)
	}

	sort.Slice(summary, func(i, j int) bool {
		return summary[i].Count > summary[j].Count
	})

	return ListResourcesResult{
		Total:   total,
		ByType:  byType,
		Summary: summary,
	}, nil
}

File: gcp_cost_optimizer_agent/python/tools/assets.py
"""Cloud Asset Inventory tool — list all GCP resources in a project."""

from __future__ import annotations

from google.cloud import asset_v1


def list_resources(project_id: str, asset_types: list[str] = []) -> dict:  # noqa: B006
    """List all GCP resources in a project, grouped by type.

    Args:
        project_id: The GCP project ID (e.g. "my-project").
        asset_types: Optional list of asset types to filter by, e.g.
            ["compute.googleapis.com/Instance", "storage.googleapis.com/Bucket"].
            If omitted, returns all resource types.

    Returns:
        A dict with:
          - "total": total number of resources found
          - "by_type": dict mapping asset_type -> list of resource names
          - "summary": list of {"asset_type": str, "count": int} sorted by count desc
    """
    client = asset_v1.AssetServiceClient()
    request = asset_v1.ListAssetsRequest(
        parent=f"projects/{project_id}",
        asset_types=asset_types,
        content_type=asset_v1.ContentType.RESOURCE,
        page_size=1000,
    )

    by_type: dict[str, list[str]] = {}
    for asset in client.list_assets(request=request):
        by_type.setdefault(asset.asset_type, []).append(asset.name)

    summary = sorted(
        [{"asset_type": t, "count": len(names)} for t, names in by_type.items()],
        key=lambda x: x["count"],
        reverse=True,
    )

    return {
        "total": sum(len(v) for v in by_type.values()),
        "by_type": by_type,
        "summary": summary,
    }
